### ISSUE TITLE
luamplib incompatible with testphase=phase-II or phase-III

### ISSUE THREAD

**Issue by mbertucci47** [latex3/tagging-project#8]
I'm not sure if the point of this repo is to compile a big list of incompatible packages or not, but if not please feel free to delete this post. Currently the [`luamplib`](https://www.ctan.org/pkg/luamplib) package does not work with `testphase=phase-II` or `phase-III`. For example
```tex
\DocumentMetadata{testphase=phase-II} % works with phase-I
\documentclass{article}
\usepackage{luamplib}

\begin{document}

\begin{mplibcode}
beginfig(1);
draw (0,0)--(100,0)--(100,100)--(0,100)--cycle;
endfig;
\end{mplibcode}

\end{document}
```
produces
```
This is LuaHBTeX, Version 1.17.0 (TeX Live 2023)
 restricted system commands enabled.
(./luamplib-test.tex
LaTeX2e <2023-06-01> patch level 1
L3 programming layer <2023-08-11>
 (c:/texlive/2023/texmf-dist/tex/latex/latex-lab/documentmetadata-support.ltx
(c:/texlive/2023/texmf-dist/tex/latex/pdfmanagement-testphase/pdfmanagement-tes
tphase.sty (c:/texlive/2023/texmf-dist/tex/latex/tagpdf/tagpdf-base.sty)
(c:/texlive/2023/texmf-dist/tex/latex/pdfmanagement-testphase/pdfmanagement-tes
tphase.ltx
(c:/texlive/2023/texmf-dist/tex/latex/l3experimental/l3bitset/l3bitset.sty))))
(c:/texlive/2023/texmf-dist/tex/latex/l3backend/l3backend-luatex.def)
(c:/texlive/2023/texmf-dist/tex/latex/pdfmanagement-testphase/l3backend-testpha
se-luatex.def
(c:/texlive/2023/texmf-dist/tex/latex/pdfmanagement-testphase/l3ref-tmp.sty))
(c:/texlive/2023/texmf-dist/tex/latex/latex-lab/phase-II-latex-lab-testphase.lt
x
(c:/texlive/2023/texmf-dist/tex/latex/latex-lab/latex-lab-testphase-new-or-1.st
y (c:/texlive/2023/texmf-dist/tex/latex/etoolbox/etoolbox.sty)
Patching header in \@outputpage
Patching footer in \@outputpage
Patching \@makecol for tagging
Patching \@makecol for tagging
)
(c:/texlive/2023/texmf-dist/tex/latex/latex-lab/latex-lab-testphase-new-or-2.st
y (c:/texlive/2023/texmf-dist/tex/latex/latex-lab/latex-lab-footnotes.ltx))
(c:/texlive/2023/texmf-dist/tex/latex/tagpdf/tagpdf.sty
(c:/texlive/2023/texmf-dist/tex/latex/tagpdf/tagpdf-luatex.def)
(c:/texlive/2023/texmf-dist/tex/latex/tagpdf/tagpdf-mc-code-lua.sty)))
(c:/texlive/2023/texmf-dist/tex/latex/pdfmanagement-testphase/pdfmanagement-fir
staid.sty) (c:/texlive/2023/texmf-dist/tex/latex/base/article.cls
Document Class: article 2023/05/17 v1.4n Standard LaTeX document class
(c:/texlive/2023/texmf-dist/tex/latex/base/size10.clo))
(c:/texlive/2023/texmf-dist/tex/luatex/luamplib/luamplib.sty)
(c:/texlive/2023/texmf-dist/tex/latex/hyperref/nameref.sty
(c:/texlive/2023/texmf-dist/tex/latex/refcount/refcount.sty
(c:/texlive/2023/texmf-dist/tex/generic/ltxcmds/ltxcmds.sty)
(c:/texlive/2023/texmf-dist/tex/generic/infwarerr/infwarerr.sty))
(c:/texlive/2023/texmf-dist/tex/generic/gettitlestring/gettitlestring.sty
(c:/texlive/2023/texmf-dist/tex/latex/kvoptions/kvoptions.sty
(c:/texlive/2023/texmf-dist/tex/latex/graphics/keyval.sty)
(c:/texlive/2023/texmf-dist/tex/latex/kvsetkeys/kvsetkeys.sty))))
(./luamplib-test.aux) (c:/texlive/2023/texmf-dist/tex/latex/base/ts1cmr.fd)
[1{c:/texlive/2023/texmf-var/fonts/map/pdftex/updmap/pdftex.map}]
(./luamplib-test.aux)

! Package tagpdf Error: The number of automatic begin (1) and end (0)
(tagpdf)                text-unit para hooks differ!

For immediate help type H <return>.
 ...

l.14 \end{document}

?
```

**Comment by u-fischer** [comment]
The problem is that luamplib uses a low-level `\vbox` but doesn't properly end with a `\par` and the para/begin and para/end are unbalanced. This is quite independent from tagging and should be corrected in luamplib (see also ltpara-doc.pdf, end of page 2):

~~~~
\documentclass{article}
\usepackage{luamplib}



\begin{document}
\AddToHook{para/begin}{begin} \AddToHook{para/end}{end}
 
\begin{mplibcode}
beginfig(1); draw (0,0)--(100,0)--(100,100)--(0,100)--cycle; endfig; 
\end{mplibcode}


\def\mplibstoptoPDF{%
  \par % missing
  \egroup %
  \setbox\mplibscratchbox\hbox %
    {\hskip-\MPllx bp%
     \raise-\MPlly bp%
     \box\mplibscratchbox}%
  \setbox\mplibscratchbox\vbox to \MPheight
    {\vfill
     \hsize\MPwidth
     \wd\mplibscratchbox0pt%
     \ht\mplibscratchbox0pt%
     \dp\mplibscratchbox0pt%
     \box\mplibscratchbox}%
  \wd\mplibscratchbox\MPwidth
  \ht\mplibscratchbox\MPheight
  \box\mplibscratchbox
  \egroup
}

\begin{mplibcode}
beginfig(1); draw (0,0)--(100,0)--(100,100)--(0,100)--cycle; endfig; 
\end{mplibcode}

\end{document}
~~~~

![image](https://github.com/latex3/tagging-project/assets/4047173/54909f3e-9e06-4fcc-aaf8-1c5ebb779ee2)

**Comment by mbertucci47** [comment]
I think this can be closed since it was fixed in [lualatex/luamplib#114](https://github.com/lualatex/luamplib/pull/114)
